tests/AMFDeserializeTests.cpp

```diff
@@ -0,0 +1,286 @@
+#include <chrono>
+#include <fstream>
+#include <iostream>
+
+#include "AMFDeserialize.h"
+#include "AMFFormat.h"
+#include "CommonCxxTests.h"
+
+AMFArrayValue* ReadFromBitStream(RakNet::BitStream* bitStream) {
+	AMFDeserialize reader;
+
+	return reader.Read(bitStream, true);
+}
+
+/**
+ * This test checks that if we recieve an unimplemented AMFValueType 
+ * we correctly throw an error and can actch it.
+ */
+void TestUnimplementedAMFValues() {
+	std::vector<AMFValueType> unimplementedValues = {
+		AMFValueType::AMFInteger,
+		AMFValueType::AMFXMLDoc,
+		AMFValueType::AMFDate,
+		AMFValueType::AMFObject,
+		AMFValueType::AMFXML,
+		AMFValueType::AMFByteArray,
+		AMFValueType::AMFVectorInt,
+		AMFValueType::AMFVectorUInt,
+		AMFValueType::AMFVectorDouble,
+		AMFValueType::AMFVectorObject,
+		AMFValueType::AMFDictionary
+		};
+	// Run unimplemented tests to check that errors are thrown if
+	// unimplemented AMF values are attempted to be parsed.
+	std::ifstream fileStream;
+	fileStream.open("AMFBitStreamUnimplementedTest.bin", std::ios::binary);
+
+	// Read a test BitStream from a file
+	std::vector<char> baseBitStream;
+	char byte = 0;
+	while (fileStream.get(byte)) {
+		baseBitStream.push_back(byte);
+	}
+
+	fileStream.close();
+
+	for (auto amfValueType : unimplementedValues) {
+		RakNet::BitStream testBitStream;
+		for (auto element : baseBitStream) {
+			testBitStream.Write(element);
+		}
+		testBitStream.Write(amfValueType);
+		bool caughtException = false;
+		try {
+			ReadFromBitStream(&testBitStream);
+		} catch (AMFValueType unimplementedValueType) {
+			caughtException = true;
+		}
+		std::cout << "Testing unimplemented value " << amfValueType << " Did we catch an exception: " << (caughtException ? "YES" : "NO") << std::endl;
+		assert(caughtException);
+	}
+}
+
+void TestLoad() {
+	// Now run a load test.
+	auto startTime = std::chrono::high_resolution_clock::now();
+	std::ifstream fileStream;
+	fileStream.open("AMFBitStreamLoadTest.bin", std::ios::binary);
+
+	// Read a test BitStream from a file
+	RakNet::BitStream bitStream;
+	char byte = 0;
+	while (fileStream.get(byte)) {
+		bitStream.Write<char>(byte);
+	}
+
+	fileStream.close();
+
+	ReadFromBitStream(&bitStream);
```

Comment:
All of these tests likely need to implement the proper memory management.  Since the AMF format we use does not properly delete stuff I did not implement any memory management since you would need to access every possible node here to delete them and thats not what the test does.  As mentioned in another reply, the delete for AMF format will be enhanced to properly manage memory when delete is called on it so it deletes all possible and existing associative maps and dense values.

Reply:
Addressed